design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,298 @@
+Support Dynamic Primary Pod NIC Name
+=
+
+# Summary
+
+This proposal proposes a mechanism to dynamically determine the primary pod interface for KuveVirt virtual machines (
+VMs).
+The goal is to move away from the hard coded `eth0` interface and allow flexible, environment-driven interface
+selection.
+
+# Overview
+
+## Motivation
+
+In KubeVirt, each VM runs as a Kubernetes pod, and these pods have network interfaces attached to different networks.
+The default primary pod interface is often named `eth0`.
+However, users might require custom network setups where the pod’s primary interface is not named `eth0`.
+Currently, KubeVirt attaches the VM’s interface destined to the pod network to the pod’s `eth0` interface.
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+## Goals
+
+- Allow the primary pod interface name to be dynamically determined based on the environment.
+- Ensure backward compatibility, so existing VMs with `eth0` as the primary interface do not break.
+- Handle the scenario of a non-eth0 primary interface with a binding plugin.
+
+## Non Goals
+
+- Change Kubernetes' API.
+- Change Multus' API.
+
+## Definition of Users
+
+- Cluster Administrator
+- VM owner
+
+## User Stories
+
+- As a cluster administrator I want to use a CNI that configures pods’ primary interface to a name other than `eth0`.
+
+- As a VM owner I want my VMs to be connected to the cluster's default network, as they do today.
+
+- As a VM owner I want a correct interface status reporting.
+
+## Repos
+
+- https://github.com/kubevirt/kubevirt
+
+# Design
```

Comment:
@ormergi commented: From what I understand, we need to pass yet another piece of information from pod annotation down to virt-handler and virt-launcher.
Given that virt-controller have access to the pod and to the subject information.
And there is a mechanism that enable passing info from virt-controller to virt-handler, that is [kubevirt.io/network-info](https://www.google.com/url?q=http://kubevirt.io/network-info&sa=D&source=docs&ust=1726418202169902&usg=AOvVaw37lOjlGKZZOW_omCD_X31s) annotation.
Can we use this mechanism to pass the primary network interface name as well?
Is so, I think it will save from extending the status API.

Reply:
@ormergi I've added your suggestion as an alternative, it will be visible in the next push.

Since the `kubevirt.io/network-info` annotation is set on the virt-launcher pod (and not on the VMI object), virt-handler doesn't have access to it.

Since this information is needed in virt-handler as well (for network setup and network status report), a new mechanism should be created in order to share this information.

Another con to this suggestion, is the need to have the downward API for every virt-launcher pod, in comparison to just the ones derived from VMIs with SR-IOV NICs or network binding plugins that require it.